annif/backend/yake.py

```diff
@@ -0,0 +1,191 @@
+"""Annif backend using Yake keyword extraction"""
+# Enabling this optional backend may result in GPLv3 terms to cover the
+# application, because YAKE (https://github.com/LIAAD/yake) is licensed under
+# GPLv3.
+
+import yake
+import os.path
+import re
+from collections import defaultdict
+from rdflib.namespace import SKOS
+import annif.util
+from . import backend
+from annif.suggestion import SubjectSuggestion, ListSuggestionResult
+from annif.exception import ConfigurationException
+
+
+class YakeBackend(backend.AnnifBackend):
+    """Yake based backend for Annif"""
+    name = "yake"
+    needs_subject_index = False
+
+    # defaults for uninitialized instances
+    _index = None
+    _graph = None
+    INDEX_FILE = 'yake-index'
+
+    DEFAULT_PARAMETERS = {
+        'max_ngram_size': 4,
+        'deduplication_threshold': 0.9,
+        'deduplication_algo': 'levs',
+        'window_size': 1,
+        'num_keywords': 100,
+        'features': None,
+        'label_types': ['prefLabel', 'altLabel'],
+        'remove_parentheses': False
+    }
+
+    def default_params(self):
+        params = backend.AnnifBackend.DEFAULT_PARAMETERS.copy()
+        params.update(self.DEFAULT_PARAMETERS)
+        return params
+
+    @property
+    def is_trained(self):
+        return True
+
+    @property
+    def label_types(self):
+        if type(self.params['label_types']) == str:  # Label types set by user
+            label_types = [lt.strip() for lt
+                           in self.params['label_types'].split(',')]
+            self._validate_label_types(label_types)
+        else:
+            label_types = self.params['label_types']  # The defaults
+        return [getattr(SKOS, lt) for lt in label_types]
+
+    def _validate_label_types(self, label_types):
+        for lt in label_types:
+            if lt not in ('prefLabel', 'altLabel', 'hiddenLabel'):
+                raise ConfigurationException(
+                    f'invalid label type {lt}', backend_id=self.backend_id)
+
+    def initialize(self):
+        self._initialize_index()
+
+    def _initialize_index(self):
+        if self._index is None:
+            path = os.path.join(self.datadir, self.INDEX_FILE)
+            if os.path.exists(path):
+                self._index = self._load_index(path)
+                self.info(
+                    f'Loaded index from {path} with {len(self._index)} labels')
+            else:
+                self.info('Creating index')
+                self._index = self._create_index()
+                self._save_index(path)
+                self.info(f'Created index with {len(self._index)} labels')
```

Comment:
Now the index is created on a suggest call only when the index does not yet exist (it takes about 1 min to create it, a bit too long to create it on every call?). With existing index, if a project's vocabulary is updated, the update does not change the Yake backend index. The user should reset the project e.g. by  `annif clear <project>` to get the vocab update to come into effect, which is a different to the way that trainable backends need (i.e. retraining).

Here could be a condition to check if the vocabulary is newer than the index file, and  recreate the index if so. But perhaps it is better to implement this with/after issue #485, as probably some common functionalities can be used.

Reply:
Related to recreating the index: if a user changes the project configuration related to index (labeltypes, `remove_parentheses` boolean) the change does not have the expected effect if the user does not also clear the project.

Reply:
I think the current code is good; this just needs to be documented on the wiki page of the backend.